go/vt/mysqlctl/xtrabackupengine.go

```diff
@@ -0,0 +1,460 @@
+
+	// first download the file into a tmp dir
+	// use the latest backup
+
+	bh := bhs[len(bhs)-1]
+
+	// extract all the files
+	if err := be.restoreFile(ctx, cnf, logger, bh, *backupStorageHook, *backupStorageCompress, be.backupFileName(), hookExtraEnv); err != nil {
+		logger.Errorf("error restoring backup file %v:%v", be.backupFileName(), err)
+		return mysql.Position{}, err
+	}
+
+	// copy / extract files
+	logger.Infof("Restore: Preparing the files")
+	// prepare the backup
+	restoreProgram := path.Join(*xtrabackupEnginePath, xtrabackup)
+	flagsToExec := []string{"--defaults-file=" + cnf.path,
+		"--prepare",
+		"--target-dir=" + cnf.TmpDir,
+	}
+	prepareCmd := exec.Command(restoreProgram, flagsToExec...)
+	prepareOut, _ := prepareCmd.StdoutPipe()
+	prepareErr, _ := prepareCmd.StderrPipe()
+	if err = prepareCmd.Start(); err != nil {
+		return mysql.Position{}, vterrors.Wrap(err, "unable to start prepare")
+	}
+
+	errOutput, _ := ioutil.ReadAll(prepareErr)
+	stdOutput, _ := ioutil.ReadAll(prepareOut)
+	err = prepareCmd.Wait()
+	if string(stdOutput) != "" {
+		// TODO remove after testing
+		logger.Infof("Prepare stdout %v", string(stdOutput))
+	}
+	output := string(errOutput)
+	if output != "" {
+		// TODO remove after testing
+		logger.Infof("Prepare stderr %v", output)
+	}
+
+	if err != nil {
+		return mysql.Position{}, err
+	}
+	if !strings.Contains(output, successMsg) {
+		return mysql.Position{}, vterrors.Errorf(vtrpc.Code_UNKNOWN, "prepare step failed: %v", output)
+	}
+
+	// then copy-back
+	logger.Infof("Restore: Copying the files")
+
+	flagsToExec = []string{"--defaults-file=" + cnf.path,
+		"--copy-back",
+		"--target-dir=" + cnf.TmpDir,
+	}
+	copybackCmd := exec.Command(restoreProgram, flagsToExec...)
+	// TODO: check stdout for OK message
+	copybackErr, _ := copybackCmd.StderrPipe()
+	copybackOut, _ := copybackCmd.StdoutPipe()
+
+	if err = copybackCmd.Start(); err != nil {
+		return mysql.Position{}, vterrors.Wrap(err, "unable to start copy-back")
+	}
+
+	errOutput, _ = ioutil.ReadAll(copybackErr)
+	stdOutput, _ = ioutil.ReadAll(copybackOut)
+	err = copybackCmd.Wait()
+	output = string(errOutput)
+	if output != "" {
+		// TODO remove after testing
+		logger.Infof("Copy-back stderr %v", string(output))
+	}
+	if string(stdOutput) != "" {
+		// TODO remove after testing
+		logger.Infof("Copy-back stdout %v", string(stdOutput))
+	}
+
+	if err != nil {
+		return mysql.Position{}, err
+	}
+	if !strings.Contains(output, successMsg) {
+		return mysql.Position{}, vterrors.Errorf(vtrpc.Code_UNKNOWN, "copy-back step failed: %v", output)
+	}
+
+	// now find the slave position and return that
+	binlogInfoFile, err := os.Open(path.Join(cnf.TmpDir, binlogInfoFileName))
+	if err != nil {
+		return mysql.Position{}, err
+	}
+	defer binlogInfoFile.Close()
+	scanner := bufio.NewScanner(binlogInfoFile)
+	scanner.Split(bufio.ScanWords)
+	counter := 0
+	var replicationPosition mysql.Position
+	for counter < 3 {
+		scanner.Scan()
+		if counter == 2 {
```

Comment:
I think we should add some context on what's going on here.

Reply:
Good point. We were parsing the `xtrabackup_binlog_info` file to get the replication position. We just found a bug in this code, so I'm going to replace it with reading the replication position from the MANIFEST, which we expect will be more robust.